compiler/noirc_evaluator/src/acir/mod.rs

```diff
@@ -91,6 +91,10 @@ struct Context<'a> {
     /// which utilizes this internal memory for ACIR generation.
     element_type_sizes_blocks: HashMap<Id<Value>, BlockId>,
 
+    /// Maps type sizes to BlockId. This is used to reuse the same BlockId if different
+    /// non-homogenous arrays end up having the same type sizes layout.
+    type_sizes_to_blocks: HashMap<Vec<usize>, BlockId>,
```

Comment:
nit: Did you consider mapping the block ID based upon (element types, length) rather than the type sizes array? It may be a smaller key as the element type sizes array grows with the length.